File: app/src/main/java/knf/kuma/backup/BackUpActivity.java
package knf.kuma.backup;

import android.animation.Animator;
import android.content.Context;
import android.content.Intent;
import android.graphics.Rect;
import android.os.Bundle;
import android.support.annotation.ColorInt;
import android.support.annotation.Nullable;
import android.support.v7.app.AppCompatActivity;
import android.util.Log;
import android.view.View;
import android.view.ViewAnimationUtils;
import android.view.animation.AccelerateDecelerateInterpolator;
import android.widget.Button;

import com.afollestad.materialdialogs.MaterialDialog;
import com.crashlytics.android.answers.Answers;
import com.crashlytics.android.answers.LoginEvent;
import com.dropbox.core.android.Auth;
import com.google.android.gms.auth.api.signin.GoogleSignIn;

import java.util.List;

import butterknife.BindView;
import butterknife.BindViews;
import butterknife.ButterKnife;
import butterknife.OnClick;
import knf.kuma.R;
import knf.kuma.custom.SyncItemView;
import xdroid.toaster.Toaster;

public class BackUpActivity extends AppCompatActivity implements BUUtils.LoginInterface, SyncItemView.OnClick {

    @BindView(R.id.colorChanger)
    View colorChanger;
    @BindView(R.id.lay_main)
    View lay_main;
    @BindView(R.id.lay_buttons)
    View lay_buttons;
    @BindViews({R.id.sync_favs, R.id.sync_history, R.id.sync_following, R.id.sync_seen})
    List<SyncItemView> syncItems;
    private boolean waitingLogin = false;

    public static void start(Context context) {
        context.startActivity(new Intent(context, BackUpActivity.class));
    }

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_login);
        ButterKnife.bind(this);
        BUUtils.init(this, savedInstanceState == null);
        if (BUUtils.isLogedIn()) {
            setState(true);
            showColor(savedInstanceState == null);
            initSyncButtons();
        } else {
            setState(false);
        }
    }

    private void initSyncButtons() {
        for (SyncItemView itemView : syncItems) {
            itemView.init(this);
        }
    }

    private void clearSyncButtons() {
        for (SyncItemView itemView : syncItems) {
            itemView.clear();
        }
    }

    @OnClick(R.id.login_dropbox)
    void onDropBoxLogin(Button button) {
        waitingLogin = true;
        BUUtils.startClient(BUUtils.BUType.DROPBOX, false);
    }

    @OnClick(R.id.login_drive)
    void onDriveLogin(Button button) {
        waitingLogin = true;
        BUUtils.startClient(BUUtils.BUType.DRIVE, false);
    }

    @Override
    public void onAction(final SyncItemView syncItemView, String id, boolean isBackup) {
        if (isBackup) {
            BUUtils.backup(id, backupObject -> {
                if (backupObject == null)
                    Toaster.toast("Error al respaldar");
                syncItemView.enableBackup(backupObject, BackUpActivity.this);
            });
        } else {
            BUUtils.restoreDialog(id, syncItemView.getBakup());
        }
    }

    @OnClick(R.id.logOut)
    void onLogOut(Button button) {
        new MaterialDialog.Builder(this)
                .content("Los datos se quedaran en el dispositivo, ¿desea continuar?")
                .positiveText("continuar")
                .negativeText("cancelar")
                .onPositive((dialog, which) -> {
                    BUUtils.logOut();
                    revertColor();
                    setState(false);
                    clearSyncButtons();
                }).build().show();
    }

    @Override
    public void onLogin() {
        if (BUUtils.isLogedIn()) {
            setState(true);
            showColor(true);
            initSyncButtons();
            Answers.getInstance().logLogin(new LoginEvent().putMethod((BUUtils.getType() == BUUtils.BUType.DRIVE) ? "Drive" : "Dropbox"));
        } else if (waitingLogin) {
            Toaster.toast("Error al iniciar sesión");
        }
        waitingLogin = false;
    }

    @ColorInt
    private int getBackColor() {
        switch (BUUtils.getType()) {
            default:
            case LOCAL:
                return getResources().getColor(android.R.color.transparent);
            case DRIVE:
                return getResources().getColor(R.color.drive);
            case DROPBOX:
                return getResources().getColor(R.color.dropbox);
        }
    }

    private void showColor(final boolean animate) {
        colorChanger.post(() -> {
            colorChanger.setBackgroundColor(getBackColor());
            if (animate) {
                Rect bounds = new Rect();
                colorChanger.getDrawingRect(bounds);
                int centerX = bounds.centerX();
                int centerY = bounds.centerY();
                int finalRadius = Math.max(bounds.width(), bounds.height());
                Animator animator = ViewAnimationUtils.createCircularReveal(colorChanger, centerX, centerY, 0f, finalRadius);
                animator.setDuration(1000);
                animator.setInterpolator(new AccelerateDecelerateInterpolator());
                colorChanger.setVisibility(View.VISIBLE);
                animator.start();
            } else {
                colorChanger.setVisibility(View.VISIBLE);
            }
        });
    }

    private void revertColor() {
        colorChanger.post(() -> {
            Rect bounds = new Rect();
            colorChanger.getDrawingRect(bounds);
            int centerX = bounds.centerX();
            int centerY = bounds.centerY();
            int finalRadius = Math.max(bounds.width(), bounds.height());
            Animator animator = ViewAnimationUtils.createCircularReveal(colorChanger, centerX, centerY, finalRadius, 0f);
            animator.setDuration(1000);
            animator.setInterpolator(new AccelerateDecelerateInterpolator());
            animator.addListener(new Animator.AnimatorListener() {
                @Override
                public void onAnimationStart(Animator animation) {

                }

                @Override
                public void onAnimationEnd(Animator animation) {
                    colorChanger.setVisibility(View.INVISIBLE);
                }

                @Override
                public void onAnimationCancel(Animator animation) {

                }

                @Override
                public void onAnimationRepeat(Animator animation) {

                }
            });
            animator.start();
        });
    }

    private void setState(final boolean isLogedIn) {
        runOnUiThread(() -> {
            lay_main.setVisibility(isLogedIn ? View.GONE : View.VISIBLE);
            lay_buttons.setVisibility(isLogedIn ? View.VISIBLE : View.GONE);
        });
    }

    @Override
    protected void onResume() {
        super.onResume();
        if (waitingLogin) {
            String token = Auth.getOAuth2Token();
            if (token != null)
                BUUtils.setType(BUUtils.BUType.DROPBOX);
            BUUtils.setDropBoxClient(token);
        }
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (requestCode == BUUtils.LOGIN_CODE) {
            Log.e("Result", "Code: " + resultCode);
            if (resultCode == RESULT_OK) {
                GoogleSignIn.getSignedInAccountFromIntent(data);
                BUUtils.setType(BUUtils.BUType.DRIVE);
                BUUtils.setDriveClient();
            } else {
                onLogin();
            }
        }
    }
}

File: app/src/main/java/knf/kuma/backup/BackUpActivity.kt
package knf.kuma.backup

import android.animation.Animator
import android.content.Context
import android.content.Intent
import android.content.pm.ActivityInfo
import android.graphics.Rect
import android.os.Bundle
import android.view.View
import android.view.ViewAnimationUtils
import android.view.animation.AccelerateDecelerateInterpolator
import androidx.annotation.ColorInt
import androidx.core.content.ContextCompat
import androidx.preference.PreferenceManager
import com.afollestad.materialdialogs.MaterialDialog
import com.dropbox.core.android.Auth
import com.google.android.gms.common.ConnectionResult
import com.google.android.gms.common.GoogleApiAvailability
import knf.kuma.BuildConfig
import knf.kuma.R
import knf.kuma.achievements.AchievementManager
import knf.kuma.backup.firestore.FirestoreManager
import knf.kuma.backup.framework.BackupService
import knf.kuma.backup.framework.DropBoxService
import knf.kuma.backup.framework.LocalService
import knf.kuma.commons.EAHelper
import knf.kuma.commons.Network
import knf.kuma.commons.PrefsUtil
import knf.kuma.commons.admFile
import knf.kuma.commons.noCrash
import knf.kuma.commons.safeShow
import knf.kuma.commons.showSnackbar
import knf.kuma.custom.GenericActivity
import knf.kuma.custom.SyncItemView
import knf.kuma.databinding.ActivityLoginBinding
import kotlin.math.max

class BackUpActivity : GenericActivity(), SyncItemView.OnClick {
    val binding by lazy { ActivityLoginBinding.inflate(layoutInflater) }
    private val syncItems: MutableList<SyncItemView> by lazy {
        with(binding.layButtons) {
            arrayListOf(syncFavs, syncHistory, syncFollowing, syncSeen, syncSeenNew)
        }
    }
    private var service: BackupService? = null
    private var waitingLogin = false

    private val backColor: Int
        @ColorInt
        get() {
            return when (Backups.type) {
                Backups.Type.NONE -> ContextCompat.getColor(this, android.R.color.transparent)
                Backups.Type.LOCAL -> ContextCompat.getColor(this, EAHelper.getThemeColorLight())
                Backups.Type.DROPBOX -> ContextCompat.getColor(this, R.color.dropbox)
                Backups.Type.FIRESTORE -> ContextCompat.getColor(this, R.color.firestore)
            }
        }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        if (!resources.getBoolean(R.bool.isTablet))
            requestedOrientation = ActivityInfo.SCREEN_ORIENTATION_PORTRAIT
        setContentView(binding.root)
        service = Backups.createService()
        binding.layMain.loginDropbox.setOnClickListener { onDropBoxLogin() }
        if (!PrefsUtil.isAdsEnabled && !BuildConfig.DEBUG && !admFile.exists() && !PrefsUtil.isSubscriptionEnabled) {
            binding.layMain.loginFirestore.isEnabled = false
            binding.layMain.adsRequired.visibility = View.VISIBLE
        } else if (PrefsUtil.isAdsEnabled && Network.isAdsBlocked && !BuildConfig.DEBUG && !admFile.exists() && !PrefsUtil.isSubscriptionEnabled) {
            binding.layMain.loginFirestore.isEnabled = false
            binding.layMain.adsRequired.text = "Anuncios bloqueados por host"
            binding.layMain.adsRequired.visibility = View.VISIBLE
        } else if (!PrefsUtil.isSecurityUpdated && PrefsUtil.spProtectionEnabled && PrefsUtil.spErrorType != null) {
            binding.layMain.loginFirestore.isEnabled = false
            binding.layMain.adsRequired.text =
                "Proveedor de seguridad no pudo ser actualizado (${PrefsUtil.spErrorType})"
            binding.layMain.adsRequired.visibility = View.VISIBLE
        }
        if (GoogleApiAvailability.getInstance().isGooglePlayServicesAvailable(this) == ConnectionResult.SUCCESS) {
            FirestoreManager.start()
            binding.layMain.loginFirestore.setOnClickListener { onFirestoreLogin() }
        } else {
            binding.layMain.loginFirestore.isEnabled = false
            binding.layMain.adsRequired.text = "Google Play Services no disponibles"
            binding.layMain.adsRequired.visibility = View.VISIBLE
        }
        binding.layMain.loginLocal.setOnClickListener { onLocalLogin() }
        binding.layButtons.logOut.setOnClickListener { onLogOut() }
        binding.layFirestore.logOutFirestore.setOnClickListener { onLogOut() }
        when {
            service?.isLoggedIn == true -> {
                setState(true)
                showColor(savedInstanceState == null)
                initSyncButtons()
            }
            FirestoreManager.isLoggedIn -> {
                setState(true)
                showColor(savedInstanceState == null)
                initFirestoreSync()
            }
            else -> setState(false)
        }
    }

    private fun initSyncButtons() {
        for (itemView in syncItems) {
            itemView.init(service, this)
        }
    }

    private fun initFirestoreSync() {
        AchievementManager.onBackup()
        binding.layFirestore.staticSyncAchievements.suscribe(this, FirestoreManager.achievementsLiveData)
        binding.layFirestore.staticSyncEA.suscribe(this, FirestoreManager.eaLiveData)
        binding.layFirestore.staticSyncFavs.suscribe(this, FirestoreManager.favsLiveData)
        binding.layFirestore.staticSyncGenres.suscribe(this, FirestoreManager.genresLiveData)
        binding.layFirestore.staticSyncHistory.suscribe(this, FirestoreManager.historyLiveData)
        binding.layFirestore.staticSyncQueue.suscribe(this, FirestoreManager.queueLiveData)
        binding.layFirestore.staticSyncSeeing.suscribe(this, FirestoreManager.seeingLiveData)
        binding.layFirestore.staticSyncSeen.suscribe(this, FirestoreManager.seenLiveData)
    }

    private fun clearSyncButtons() {
        for (itemView in syncItems) {
            itemView.clear()
        }
    }

    private fun onDropBoxLogin() {
        waitingLogin = true
        service = DropBoxService().also { it.logIn() }
    }

    private fun onFirestoreLogin() {
        FirestoreManager.doLogin(this)
    }

    private fun onLocalLogin() {
        MaterialDialog(this).safeShow {
            message(text = "Los datos se quedarán en la memoria, no se podrá sincronizar datos entre dispositivos, usar este método?")
            positiveButton(text = "usar") {
                service = LocalService().also {
                    it.start()
                    it.logIn()
                }
                onLogin()
            }
            negativeButton(text = "cancelar")
        }
    }

    override fun onAction(syncItemView: SyncItemView, id: String, isBackup: Boolean) {
        noCrash {
            if (isBackup)
                Backups.backup(binding.colorChanger, service, id) {
                    noCrash {
                        if (it == null)
                            binding.colorChanger.showSnackbar("Error al respaldar")
                        syncItemView.enableBackup(it, this@BackUpActivity)
                    }
                }
            else
                Backups.restoreDialog(this, binding.colorChanger, id, syncItemView.backupObj)
        }
    }

    private fun onLogOut() {
        MaterialDialog(this).safeShow {
            message(text = "Los datos no respaldados podrian ser perdidos al borrar la app, ¿desea continuar?")
            positiveButton(text = "continuar") {
                if (Backups.type == Backups.Type.FIRESTORE) {
                    FirestoreManager.doSignOut(this@BackUpActivity)
                    Backups.type = Backups.Type.NONE
                    revertColor()
                    setState(false)
                } else {
                    PreferenceManager.getDefaultSharedPreferences(this@BackUpActivity).edit().putString("auto_backup", "0").apply()
                    service?.logOut()
                    service = null
                    Backups.type = Backups.Type.NONE
                    revertColor()
                    setState(false)
                    clearSyncButtons()
                }
            }
            negativeButton(text = "cancelar")
        }
    }

    private fun onLogin() {
        if (service?.isLoggedIn == true || FirestoreManager.isLoggedIn) {
            setState(true)
            showColor(true)
            initSyncButtons()
        } else if (waitingLogin) {
            binding.colorChanger.showSnackbar("Error al iniciar sesión")
        }
        waitingLogin = false
    }

    private fun showColor(animate: Boolean) {
        binding.colorChanger.post {
            try {
                binding.colorChanger.setBackgroundColor(backColor)
                if (animate) {
                    val bounds = Rect()
                    binding.colorChanger.getDrawingRect(bounds)
                    val centerX = bounds.centerX()
                    val centerY = bounds.centerY()
                    val finalRadius = max(bounds.width(), bounds.height())
                    val animator = ViewAnimationUtils.createCircularReveal(binding.colorChanger, centerX, centerY, 0f, finalRadius.toFloat())
                    animator.duration = 1000
                    animator.interpolator = AccelerateDecelerateInterpolator()
                    binding.colorChanger.visibility = View.VISIBLE
                    animator.start()
                } else {
                    binding.colorChanger.visibility = View.VISIBLE
                }
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }
    }

    private fun revertColor() {
        binding.colorChanger.post {
            val bounds = Rect()
            binding.colorChanger.getDrawingRect(bounds)
            val centerX = bounds.centerX()
            val centerY = bounds.centerY()
            val finalRadius = max(bounds.width(), bounds.height())
            val animator = ViewAnimationUtils.createCircularReveal(binding.colorChanger, centerX, centerY, finalRadius.toFloat(), 0f)
            animator.duration = 1000
            animator.interpolator = AccelerateDecelerateInterpolator()
            animator.addListener(object : Animator.AnimatorListener {
                override fun onAnimationStart(animation: Animator) {

                }

                override fun onAnimationEnd(animation: Animator) {
                    binding.colorChanger.visibility = View.INVISIBLE
                }

                override fun onAnimationCancel(animation: Animator) {

                }

                override fun onAnimationRepeat(animation: Animator) {

                }
            })
            animator.start()
        }
    }

    private fun setState(isLoggedIn: Boolean) {
        runOnUiThread {
            binding.layMain.root.visibility = if (isLoggedIn) View.GONE else View.VISIBLE
            when (Backups.type) {
                Backups.Type.LOCAL, Backups.Type.DROPBOX -> {
                    binding.layButtons.root.visibility = if (isLoggedIn) View.VISIBLE else View.GONE
                    binding.layFirestore.root.visibility = View.GONE
                }
                Backups.Type.FIRESTORE -> {
                    binding.layFirestore.root.visibility = if (isLoggedIn) View.VISIBLE else View.GONE
                    binding.layButtons.root.visibility = View.GONE
                }
                Backups.Type.NONE -> {
                    binding.layButtons.root.visibility = if (isLoggedIn) View.VISIBLE else View.GONE
                    binding.layFirestore.root.visibility = if (isLoggedIn) View.VISIBLE else View.GONE
                }
            }
        }
    }

    override fun onResume() {
        super.onResume()
        if (waitingLogin) {
            val token = Auth.getOAuth2Token()
            if (service is DropBoxService && service?.logIn(token) == true) {
                Backups.type = Backups.Type.DROPBOX
            }
            onLogin()
        }
    }


    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (FirestoreManager.handleLogin(this, requestCode, resultCode, data)) {
            initFirestoreSync()
        }
        onLogin()
    }

    companion object {

        fun start(context: Context) {
            context.startActivity(Intent(context, BackUpActivity::class.java))
        }
    }
}
